File: matiasmasso/Mat.NET/DTO/DTOZip.cs
using MatHelperStd;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOZip : DTOArea
    {
        public string ZipCod { get; set; }
        public DTOLocation Location { get; set; }
        public List<DTOContact> Contacts { get; set; }

        public DTOZip() : base()
        {
            base.Cod = Cods.Zip;
        }

        public DTOZip(Guid oGuid) : base(oGuid)
        {
            base.Cod = Cods.Zip;
        }

        public static DTOZip Factory(DTOLocation oLocation, string sZipCod)
        {
            DTOZip retval = new DTOZip();
            {
                var withBlock = retval;
                withBlock.Location = oLocation;
                withBlock.ZipCod = sZipCod;
            }
            return retval;
        }

        public string ZipyCit()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (this.ZipCod.isNotEmpty())
                sb.Append(this.ZipCod + " ");
            if (this.Location != null)
                sb.Append(this.Location.Nom);
            string retval = sb.ToString();
            return retval;
        }

        public static string ZipyCit(DTOZip oZip)
        {
            string retval = "";
            if (oZip != null)
                retval = oZip.ZipyCit();
            return retval;
        }

        public string FullNom(DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (this.ZipCod.isNotEmpty())
                sb.Append(this.ZipCod);
            if (this.Location != null)
            {
                sb.Append(" ");
                sb.Append(this.Location.FullNom(oLang));
            }

            string retval = sb.ToString();
            return retval;
        }

        public static string FullNom(DTOZip oZip, DTOLang oLang = null/* TODO Change to default(_) if this is not a reference type */)
        {
            string retval = "";
            if (oZip != null)
                retval = oZip.FullNom(oLang);
            return retval;
        }

        public static string FullNomSegmented(DTOZip oZip, DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append(oZip.ZipyCit());
            sb.Append("/");
            sb.Append(oZip.Location.Zona.FullNomSegmented(oLang));
            string retval = sb.ToString();
            return retval;
        }

        public static string FullNomSegmentedReversed(DTOZip oZip, DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append(DTOLocation.FullNomSegmented(oZip.Location, oLang));
            sb.Append("/");
            sb.Append(oZip.ZipCod);
            string retval = sb.ToString();
            return retval;
        }

        public bool Matches(string searchKey)
        {
            return MatHelperStd.TextHelper.Match(this.ZipCod, searchKey);
        }

        public static bool Validate(DTOCountry oCountry, string sZipCod)
        {
            bool retval = true;
            if (oCountry.Equals(DTOCountry.Wellknown(DTOCountry.Wellknowns.Spain)))
            {
                var sPattern = "^[0-9]{5}$";
                retval = TextHelper.RegexMatch(sZipCod, sPattern);
            }
            else if (oCountry.Equals(DTOCountry.Wellknown(DTOCountry.Wellknowns.Portugal)))
            {
                var sPattern = "^[0-9]{4}-[0-9]{3}$";
                retval = TextHelper.RegexMatch(sZipCod, sPattern);
            }
            return retval;
        }

        public string CountryNom(DTOLang oLang)
        {
            string retval = "";
            if (this.Location != null)
                retval = this.Location.CountryNom(oLang);
            return retval;
        }

        public DTOInvoice.ExportCods ExportCod()
        {
            return ExportCod(this);
        }

        public static DTOInvoice.ExportCods ExportCod(DTOZip oZip)
        {
            DTOInvoice.ExportCods retval = DTOInvoice.ExportCods.notSet;
            if (oZip != null)
            {
                if (oZip.Location != null)
                    retval = DTOLocation.ExportCod(oZip.Location);
            }
            return retval;
        }

        public static DTOAreaProvincia Provincia(DTOZip oZip)
        {
            DTOAreaProvincia retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oZip != null)
            {
                if (oZip.Location != null)
                    retval = DTOLocation.Provincia(oZip.Location);
            }
            return retval;
        }



        public static List<DTOCountry> countries(IEnumerable<DTOZip> oZips)
        {
            var retval = oZips.GroupBy(x => x.Location.Zona.Country.Guid).Select(y => y.First()).Select(z => z.Location.Zona.Country).ToList();
            return retval;
        }

        public static List<DTOZona> Zonas(IEnumerable<DTOZip> oZips, DTOCountry FromCountry = null/* TODO Change to default(_) if this is not a reference type */)
        {
            var retval = oZips.GroupBy(x => x.Location.Zona.Guid).Select(y => y.First()).Select(z => z.Location.Zona).ToList();
            if (FromCountry != null)
                retval = retval.Where(x => x.Country.Equals(FromCountry)).ToList();
            return retval;
        }

        public static List<DTOLocation> Locations(IEnumerable<DTOZip> oZips, DTOZona FromZona = null/* TODO Change to default(_) if this is not a reference type */)
        {
            var retval = oZips.GroupBy(x => x.Location.Guid).Select(y => y.First()).Select(z => z.Location).ToList();
            if (FromZona != null)
                retval = retval.Where(x => x.Zona.Equals(FromZona)).ToList();
            return retval;
        }

        public static List<DTOZip> Zips(IEnumerable<DTOZip> oZips, DTOLocation FromLocation)
        {
            var retval = oZips.Where(x => x.Location.Equals(FromLocation)).ToList();
            return retval;
        }

        public class Collection : List<DTOZip>
        {

        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOZip.vb
Public Class DTOZip
    Inherits DTOArea

    Property zipCod As String
    Shadows Property location As DTOLocation
    Property contacts As List(Of DTOContact)

    Public Sub New()
        MyBase.New()
        MyBase.Cod = Cods.Country
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        MyBase.Cod = Cods.Country
    End Sub

    Shared Shadows Function Factory(oLocation As DTOLocation, sZipCod As String) As DTOZip
        Dim retval As New DTOZip
        With retval
            .Location = oLocation
            .ZipCod = sZipCod
        End With
        Return retval
    End Function

    Public Function ZipyCit() As String
        Dim sb As New System.Text.StringBuilder
        If _ZipCod > "" Then
            sb.Append(_ZipCod & " ")
        End If
        If _Location IsNot Nothing Then
            sb.Append(_Location.Nom)
        End If
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Shared Function ZipyCit(oZip As DTOZip) As String
        Dim retval As String = ""
        If oZip IsNot Nothing Then
            retval = oZip.ZipyCit
        End If
        Return retval
    End Function

    Public Shadows Function FullNom(oLang As DTOLang) As String
        Dim sb As New System.Text.StringBuilder
        If _ZipCod > "" Then
            sb.Append(_ZipCod)
        End If
        If _Location IsNot Nothing Then
            sb.Append(" ")
            sb.Append(_Location.FullNom(oLang))
        End If

        Dim retval As String = sb.ToString
        Return retval
    End Function

  shadows  Shared Function FullNom(oZip As DTOZip, Optional oLang As DTOLang = Nothing) As String
        Dim retval As String = ""
        If oZip IsNot Nothing Then
            retval = oZip.FullNom(oLang)
        End If
        Return retval
    End Function

    Shared Shadows Function FullNomSegmented(oZip As DTOZip, oLang As DTOLang) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(oZip.ZipyCit())
        sb.Append("/")
        sb.Append(oZip.Location.Zona.FullNomSegmented(oLang))
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Shared Shadows Function FullNomSegmentedReversed(oZip As DTOZip, oLang As DTOLang) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(DTOLocation.FullNomSegmented(oZip.Location, oLang))
        sb.Append("/")
        sb.Append(oZip.ZipCod)
        Dim retval As String = sb.ToString
        Return retval
    End Function


    Shared Function Validate(oCountry As DTOCountry, sZipCod As String) As Boolean
        Dim retval As Boolean = True
        If oCountry.Equals(DTOCountry.wellknown(DTOCountry.wellknowns.Spain)) Then
            Dim sPattern = "^[0-9]{5}$"
            retval = TextHelper.RegexMatch(sZipCod, sPattern)
        ElseIf oCountry.Equals(DTOCountry.wellknown(DTOCountry.wellknowns.Portugal)) Then
            Dim sPattern = "^[0-9]{4}-[0-9]{3}$"
            retval = TextHelper.RegexMatch(sZipCod, sPattern)
        End If
        Return retval
    End Function

    Public Function CountryNom(oLang As DTOLang) As String
        Dim retval As String = ""
        If _Location IsNot Nothing Then
            retval = _Location.CountryNom(oLang)
        End If
        Return retval
    End Function

    Shared Function ExportCod(oZip As DTOZip) As DTOInvoice.ExportCods
        Dim retval As DTOInvoice.ExportCods = DTOInvoice.ExportCods.NotSet
        If oZip IsNot Nothing Then
            If oZip.Location IsNot Nothing Then
                retval = DTOLocation.ExportCod(oZip.Location)
            End If
        End If
        Return retval
    End Function

    Shared Shadows Function Provincia(oZip As DTOZip) As DTOAreaProvincia
        Dim retval As DTOAreaProvincia = Nothing
        If oZip IsNot Nothing Then
            If oZip.Location IsNot Nothing Then
                retval = DTOLocation.Provincia(oZip.Location)
            End If
        End If
        Return retval
    End Function



    Shared Function Countries(oZips As IEnumerable(Of DTOZip)) As List(Of DTOCountry)
        Dim retval = oZips.GroupBy(Function(x) x.Location.Zona.Country.Guid).Select(Function(y) y.First).Select(Function(z) z.Location.Zona.Country).ToList
        Return retval
    End Function

    Shared Function Zonas(oZips As IEnumerable(Of DTOZip), Optional FromCountry As DTOCountry = Nothing) As List(Of DTOZona)
        Dim retval = oZips.GroupBy(Function(x) x.Location.Zona.Guid).Select(Function(y) y.First).Select(Function(z) z.Location.Zona).ToList
        If FromCountry IsNot Nothing Then
            retval = retval.Where(Function(x) x.Country.Equals(FromCountry)).ToList
        End If
        Return retval
    End Function

    Shared Function Locations(oZips As IEnumerable(Of DTOZip), Optional FromZona As DTOZona = Nothing) As List(Of DTOLocation)
        Dim retval = oZips.GroupBy(Function(x) x.Location.Guid).Select(Function(y) y.First).Select(Function(z) z.Location).ToList
        If FromZona IsNot Nothing Then
            retval = retval.Where(Function(x) x.Zona.Equals(FromZona)).ToList
        End If
        Return retval
    End Function

    Shared Function Zips(oZips As IEnumerable(Of DTOZip), FromLocation As DTOLocation) As List(Of DTOZip)
        Dim retval = oZips.Where(Function(x) x.Location.Equals(FromLocation)).ToList
        Return retval
    End Function
End Class